### DISCUSSION TITLE
Next 9.5 monorepo production issue

### DISCUSSION THREAD

**Discussion by stevez86** [vercel/next.js#23981] [Help]
# Bug report

After updating to 9.5, next no longer works in my current monorepo setup on a deployment to vercel. Everything appears fine in development, however once I deploy to vercel the issue is exhibited by the home page redirecting to a path identical to the routing setup to the routes field in the now.json. In this case, the homepage is redirected to /my-app

Here is a link to the reproduction: https://github.com/stevez86/next-repro-2
And a link to a deployment: https://my-app-nsgjc5j8l.vercel.app/

Here is my now.json file, which I think is the culprit. The other app routes appear to be working as intended ie /abc/:slug1/:slug2 so I think this is only applicable to the homepage route at "/". 

```javascript
{
  "version": 2,
  "builds": [{ "src": "my-app/next.config.js", "use": "@now/next" }],
  "routes": [{ "src": "/(.*)", "dest": "/my-app/$1", "continue": true }]
}
```

**Comment by fredrik-sogaard** [comment]
We're seeing this as well. Similar setup.

**Comment by fredrik-sogaard** [comment]
I removed `"continue": true` and it seems to work, but it's very weird issue and sometimes it seems to linger in browser cache. Will report back if this didn't solve it.

**Comment by fredrik-sogaard** [comment]
Turn out that removing `"continue": true` introduced new issues, so we simply solved this better by moving away from `routes` to `rewrites` in our now.json file.

**Comment by stevez86** [comment]
@fredrik-sogaard Interesting. Mind providing an example?

**Comment by GxDesign** [comment]
Same, an example would be great. Im having issues with being able to use a catch all, its overriding my other rewrites

**Comment by fredrik-sogaard** [comment]
This worked for us:
```
  "rewrites": [
    {
      "source": "/backend",
      "destination": "packages/backend/src/index.ts"
    },
    {
      "source": "/(.*)",
      "destination": "packages/frontend/$1"
    }
  ],

```

**Comment by stevez86** [comment]
@timneutkens Is this an issue with Next or Now? Should I be doing monorepo routing differently or is the rewrites method the way we should be handling this?

**Comment by fredrik-sogaard** [comment]
From what I could gather, rewrites is the way forward.

We only encountered the bug when upgrading from Next 9.4 to 9.5 and deployed to Vercel, so I'm guessing it's both a Vercel and Next issue?

**Comment by GxDesign** [comment]
Are you able to run that with `vercel dev`? For me, that "/(.*)" will catch all the routes including /backend

**Comment by fredrik-sogaard** [comment]
Haven't tried with `vercel dev` but we're running this setup on a production site now on Vercel. 

Can you share your now.json file?

**Comment by GxDesign** [comment]
Yes, thanks for helping:

```
{
  "version": 2,
  "build": {
    "env": {
      "BUILDING_FOR_NOW": "true"
    }
  },
  "builds": [
    { "src": "packages/clubs/package.json", "use": "@vercel/next" },
    { "src": "packages/books/package.json", "use": "@vercel/next" }
  ],
  "rewrites": [
    { "source": "/books(.*)", "destination": "/packages/books$1" },
    { "source": "/(.*)", "destination": "/packages/clubs/$1" }
  ]
}

```
If I remove the second rewrite, the first one works. The app follows the multi-zones setup but im trying to use rewrites instead of routes and I have two nextjs apps under a packages directory.

**Comment by GxDesign** [comment]
Using a negative lookahead also causes the first route to be rewritten to the wrong package: 

`{ "source": "/((?!books).*)", "destination": "/packages/clubs/$1" }`

**Comment by GxDesign** [comment]
I figured it out:
{ "source": "(/(?!packages).*)", "destination": "/packages/book-clubs$1" }

The previous routes were rewritten to /packages, so the catch all route needs to have a negative lookahead to not catch the previous rewrites.

**Comment by samcx** [comment]
Hello,

We don't recommend attempting to make monorepos work via a `vercel.json` file — it's simply a pain attempting to make the routing work.

We are testing Monorepo Support internally without the need for a `vercel.json` file. Stay tuned for an announcement in the coming days!

**Comment by AntonioRedondo** [comment]
We're having the same issue after updating to Next.js 9.5. But we don't use Vercel for deployments. It works fine on local dev mode. It breaks on home showing 404 once we deploy in our servers. Which may mean the bug happens on `next build` or `next start`?

**Comment by stevez86** [comment]
@samsisle Very Interesting! Any rough idea when that example will be released?

**Comment by samcx** [comment]
@stevez86 It was released! → https://vercel.com/blog/monorepos 🎉

**Comment by stevez86** [comment]
I just migrated to the recommended way of using monorepos and I'm still seeing the issue

**Reply by mcsdevv** [reply]
Hi @stevez86, I'm sorry to hear you're having trouble.

Please create a case [here](https://vercel.com/support/request) with all the relevant details and reproduction so that the Vercel Support team can investigate this for you.
